### PR TITLE
Adding parser support for N-d arrays up to rank 32

### PR DESCRIPTION
**Pull request by T-Gro**
This PR extends the parser to support definition of types for multidimensional arrays.
Until now, only backtick syntax (e.g.  int ``[,,,,]`` ) was supported.

Within this PR, backticks are no longer needed.
- The backtick syntax int ``[,,,,]``  
- syntax without backticks  int [,,,,]
- shorthand syntax array5d

Are the same type and are all supported.

The limitation of max. rank 32 is inherited by the limitation of System.Array, and the compiler will not allow int[.. 32 or more commas] as a type.

### CODE REVIEW

File: tests/FSharp.Compiler.ComponentTests/Language/MultiDimensionalArrayTests.fs [hunk 1]
```diff
@@ -0,0 +1,45 @@
+﻿namespace FSharp.Compiler.ComponentTests.Language
+
+open Xunit
+open FSharp.Test.Compiler
+
+module MultiDimensionalArrayTests =
+
+    [<Theory>]
+    [<InlineData(1,"array2d")>]
```
Comment (0101) [inline tests/FSharp.Compiler.ComponentTests/Language/MultiDimensionalArrayTests.fs:9]: I think there should be spaces after the commas. Maybe just run `fantomas` on the file.

File: tests/FSharp.Compiler.ComponentTests/Language/MultiDimensionalArrayTests.fs [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    let ``Multidimensional array - reports an error if types are not matching`` () = 
+        let commaString = System.String(',',30)
+        FSharp $"""
+module MultiDimArrayErrorTests
+let cleanStyle : int [{commaString}] = Unchecked.defaultof<_>
+let shortCutStyle : int array32d = cleanStyle
+        """
+        |> compile
+        |> shouldFail
+        |> withDiagnosticMessageMatches """This expression was expected to have type.*"""
```
Comment (0101) [inline tests/FSharp.Compiler.ComponentTests/Language/MultiDimensionalArrayTests.fs:33]: Might be worth it to spell out the full error message here?

**Review by 0101 — APPROVED** [review]

File: tests/FSharp.Compiler.ComponentTests/Language/MultiDimensionalArrayTests.fs (hunk 1 above)
Comment (T-Gro) [reply tests/FSharp.Compiler.ComponentTests/Language/MultiDimensionalArrayTests.fs:9]: Fantomas is disabled for /tests folder for all invocations.
Reformatted with ignore files turned off, should be OK now.

File: tests/FSharp.Compiler.ComponentTests/Language/MultiDimensionalArrayTests.fs (hunk 2 above)
Comment (T-Gro) [reply tests/FSharp.Compiler.ComponentTests/Language/MultiDimensionalArrayTests.fs:33]: Can get flaky/instable due to whitespace and newline, which is what I wanted avoid.
(= purpose of **this** test isnt about newlines nor formatting).

But will do, let's see how stable this is.

**Comment by T-Gro** [comment]
@dsyme can I merge this?

**Review by 0101 — APPROVED** [review]

File: tests/FSharp.Compiler.ComponentTests/FSharp.Compiler.ComponentTests.fsproj [hunk 3]
```diff
@@ -161,6 +161,7 @@
     <Compile Include="ErrorMessages\WarnIfDiscardedInList.fs" />
 	<Compile Include="ErrorMessages\UnionCasePatternMatchingErrors.fs"/>
     <Compile Include="Language\IndexerSetterParamArray.fs" />
+	<Compile Include="Language\MultiDimensionalArrayTests.fs" />
```
Comment (psfinaki) [inline tests/FSharp.Compiler.ComponentTests/FSharp.Compiler.ComponentTests.fsproj:164]: np: extra tab/spacing here

File: tests/FSharp.Compiler.ComponentTests/Language/MultiDimensionalArrayTests.fs [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+open Xunit
+open FSharp.Test.Compiler
+
+module MultiDimensionalArrayTests =
+
+    [<Theory>]
+    [<InlineData(1, "array2d")>]
+    [<InlineData(5, "array6d")>]
+    [<InlineData(31, "array32d")>]
+    let ``MultiDimensional array type can be written with or without backticks`` (commas: int, shortcut: string) =
```
Comment (psfinaki) [inline tests/FSharp.Compiler.ComponentTests/Language/MultiDimensionalArrayTests.fs:12]: So I guess it's multidimensional, backtick, and shortcut - all are one word so no need for extra capitals ;)

**Review by psfinaki — APPROVED** [review]
Feel free to ignore all the nitpicking - great job :)

**Pull request merged**